<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Reproducible environments for integrated computational workflows</title>
    <meta charset="utf-8" />
    <meta name="author" content="Kevin Rue-Albrecht and Charlotte George" />
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/metropolis.css" rel="stylesheet" />
    <link href="libs/remark-css/rladies-fonts.css" rel="stylesheet" />
    <link href="libs/font-awesome/css/all.css" rel="stylesheet" />
    <link href="libs/font-awesome/css/v4-shims.css" rel="stylesheet" />
    <link rel="stylesheet" href="my-theme.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Reproducible environments for integrated computational workflows
## Bioconductor 2020 - Birds of a Feather (BOF)
### <a href="https://kevinrue.github.io">Kevin Rue-Albrecht</a> and <a href="https://www.rdm.ox.ac.uk/people/charlotte-george">Charlotte George</a>
### University of Oxford
### 2020-07-31 (updated: 2020-07-31)

---


layout: true

&lt;div class="my-header"&gt;&lt;img src="img/BioC2020.png" alt="Bioconductor logo" align="right" height="90%"&gt;&lt;/div&gt;

&lt;div class="my-footer"&gt;&lt;span&gt;
Bioconductor 2020
&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;
Reproducible environments for integrated computational workflows
&lt;/span&gt;&lt;/div&gt;







---

# Structure of the session

.pull-left[
## What

&lt;i class="far fa-clock"&gt;&lt;/i&gt;
This session is 55 minutes.

&lt;i class="fas fa-info-circle"&gt;&lt;/i&gt;
A handful of introduction slides.

  + Workflow managers
  
  + Environment managers

&lt;i class="fas fa-volume-up"&gt;&lt;/i&gt;
Open discussion

  + Your preferences
  
  + Opinionated best practices
  
  + Feature requests
]

.pull-right[
## How

&lt;i class="far fa-comment-alt"&gt;&lt;/i&gt;
[sli.do/mrqa5tmx](http://sli.do/mrqa5tmx)

&lt;i class="fas fa-microphone"&gt;&lt;/i&gt;
Take the mic!

&lt;i class="fas fa-chart-bar"&gt;&lt;/i&gt;
Polls (on the right)

&lt;i class="fab fa-google-drive"&gt;&lt;/i&gt;
Take notes on this collaborative
[Google Docs](https://docs.google.com/document/d/1ok2H9HVoa8c-m-VcwdQM0ccOuMTmESKCzaaWHb6trwE/edit?usp=sharing)
during (and after) the session!

&lt;i class="fab fa-slack"&gt;&lt;/i&gt;
Continue the discussion on the [#reproducibility](https://community-bioc.slack.com/archives/C0185MUQ8TA) channel
in the [Bioconductor](community-bioc.slack.com) Slack workspace!

&lt;i class="fab fa-github"&gt;&lt;/i&gt;
Contribute to the [gitbook](https://github.com/kevinrue/ORCA)!
]

---

# Reproducible workflows and environments

## Why do we care?

&gt; The [bio.tools](https://bio.tools/) registry of life science software tools includes over 17,000 entries from more than 1,000 contributors &lt;a name=cite-mark2020pipelines&gt;&lt;/a&gt;([Marx, 2020](https://www.ncbi.nlm.nih.gov/pubmed/32601427)).

- We all write workflows that combine many of those tools to process and analyze data.

  + e.g., ChIP-seq, scRNA-seq, RNA-seq, proteomics
  
  + Some are published, others are "in-house"

- For full reproducibility, collaboration and teaching relies on the capacity to share:

  + Code
  
  + Environment(s): e.g., software dependencies, versions
  
  + Input data: e.g., raw data, annotations

---

# Workflow managers

.pull-left[
&lt;a href="https://www.nextflow.io/"&gt;&lt;img src="https://www.nextflow.io/img/nextflow2014_no-bg.png", height="30px", align="middle"&gt;&lt;/a&gt;
&lt;a href="https://nf-co.re/"&gt;&lt;img src="https://nf-co.re/assets/img/logo/nf-core-logo.svg", height="40px" align="middle"&gt;&lt;/a&gt;

&lt;a href="https://snakemake.readthedocs.io/en/stable/#"&gt;&lt;img src="https://learning.cyverse.org/projects/container_camp_workshop_2019/en/latest/_images/snakemake.png", height="40px" align="middle"&gt;&lt;/a&gt;

&lt;a href="http://www.ruffus.org.uk/"&gt;&lt;img src="http://www.ruffus.org.uk/_images/logo.jpg", height="50px" align="middle"&gt;&lt;/a&gt;
&lt;a href="https://github.com/cgat-developers"&gt;&lt;img src="https://avatars3.githubusercontent.com/u/5339854?s=200&amp;v=4", height="50px" align="middle"&gt;&lt;/a&gt;
&lt;a href="https://github.com/cgat-developers/cgat-core"&gt;&lt;img src="https://github.com/cgat-developers/cgat-core/raw/master/docs/img/CGAT_logo.png", height="50px" align="middle"&gt;&lt;/a&gt;

&lt;a href="https://bcbio-nextgen.readthedocs.io/en/latest/"&gt;&lt;img src="https://bcbio-nextgen.readthedocs.io/en/latest/_images/banner.png", height="50px", align="middle"&gt;&lt;/a&gt;
&lt;a href="https://workflowhub.eu/"&gt;&lt;img src="https://workflowhub.eu/assets/logos/workflowhub-20f8594d47fa27784d6a3c24010cf728e6b47753448dfa0cb6c7b7cc055c08bf.svg", height="50px", align="middle"&gt;&lt;/a&gt;

&lt;a href="https://bioconductor.org/packages/systemPipeR/"&gt;&lt;img src="https://github.com/Bioconductor/BiocStickers/blob/master/systemPipeR/systemPipeR.png?raw=true", height="80px" align="middle"&gt;&lt;/a&gt;
&lt;a href="https://bioconductor.org/packages/scPipe/"&gt;&lt;img src="https://github.com/Bioconductor/BiocStickers/blob/master/scPipe/scPipe.png?raw=true", height="80px" align="middle"&gt;&lt;/a&gt;
&lt;a href="https://docs.ropensci.org/drake/"&gt;&lt;img src="https://docs.ropensci.org/drake/reference/figures/logo.svg", height="80px" align="middle"&gt;&lt;/a&gt;

## No logo &lt;i class="far fa-sad-tear"&gt;&lt;/i&gt; ...

*[Rcwl](https://bioconductor.org/packages/3.11/Rcwl)*
]

.pull-right[
## Features

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Transform, merge, split, ... files

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Checkpoints, logs, reports, visualization

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Modular, flexible, partial runs

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; HPC queue managers and [DRMAA](http://www.drmaa.org/)

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Parallelisation

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Portable, containerized, versioned

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Different environments for some steps

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Integration with &lt;i class="fab fa-r-project"&gt;&lt;/i&gt;, &lt;i class="fab fa-python"&gt;&lt;/i&gt;, &lt;i class="fab fa-java"&gt;&lt;/i&gt;, ...
]

.center[
What is _your_ favorite?
]

---

# Environment managers

`$$Same\ workflow\ \times\ Different\ environment\ =\ Different\ result$$`
&lt;br/&gt;
&lt;br/&gt;

&lt;img src="https://i.imgur.com/k5B1CBd.jpg" height="300px" style="display: block; margin: auto;" /&gt;

---

# Environment managers

.pull-left[
&lt;a href="https://docs.conda.io/en/latest/"&gt;&lt;img src="https://docs.conda.io/en/latest/_images/conda_logo.svg", height="30px", align="middle"&gt;&lt;/a&gt;
&lt;a href="https://github.com/TheSnakePit/mamba"&gt;&lt;img src="https://miro.medium.com/max/500/1*akFPXZ_b27Bz38cnQHO3yw.png", height="40px", align="middle"&gt;&lt;/a&gt;

&lt;a href="https://www.docker.com/"&gt;&lt;img src="https://bit2bitamericas.com/wp-content/uploads/2020/01/logo-docker.png", height="30px", align="middle"&gt;&lt;/a&gt;
&lt;a href="https://sylabs.io/docs/"&gt;&lt;img src="https://sylabs.io/assets/images/logos/singularity.png", height="50px", align="middle"&gt;&lt;/a&gt;

&lt;a href="https://rstudio.github.io/renv/"&gt;&lt;img src="https://rstudio.github.io/renv/reference/figures/logo.svg", height="80px", align="middle"&gt;&lt;/a&gt;

## No logo &lt;i class="far fa-sad-tear"&gt;&lt;/i&gt; ...

*[BiocManager](https://bioconductor.org/packages/3.11/BiocManager)*
]

.pull-right[
## Features

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Track software versions

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Track source software repositories

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Portable, versioned

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Check compatibility with other packages

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Support multiple programming languages

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Track / Manage system dependencies

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Upgrade and freeze

&lt;i class="far fa-thumbs-up"&gt;&lt;/i&gt; Lightweight, fast (e.g., cache)
]

.center[
What is _your_ favorite?
]

---

# Use case: Snakemake

.pull-left[
```python
rule bwa:
    input:
        "genome.fa"
        "reads.fq"
    output:
        "mapped.bam"
    singularity:
        "docker://somecontainer:v1.0"
    conda:
        "envs/bwa.yaml"
    envmodules:
        "bio/bwa/0.7.9",
        "bio/samtools/1.9"
    shell:
        "bwa mem {input} |"
        "samtools view -Sbh - &gt; {output}"
```
]

.pull-right[
## Usage

```bash
snakemake --use-conda
```

```bash
snakemake --use-singularity
```

```bash
snakemake --use-conda --use-singularity
```

```bash
snakemake --use-envmodules
```

```bash
snakemake --use-singularity --kubernetes
```
]

---

# Paradox of choice

- What do you choose?

- How do you choose?

  + Keep sight of your objective:
    medical research?
    teaching?
    method development?

  + "Best" practice may depend on your environment and goals

  + What works for you? What can your sysadmin support?

- The community may benefit from a resource (e.g., review, book) based on experience to provide guidance and promote adoption of workflow/environment managers adapted to various context.

- Share or link (minimal) working examples for others to use as template 

---

# Discussion points

.pull-left[
## Past &amp; Present

&lt;i class="fas fa-question"&gt;&lt;/i&gt; Workflow managers

&lt;i class="fas fa-question"&gt;&lt;/i&gt; Environment managers

&lt;i class="fas fa-question"&gt;&lt;/i&gt; Challenges

&lt;i class="fas fa-question"&gt;&lt;/i&gt; Opinionated best practices
]

.pull-right[
## Future

&lt;i class="fas fa-question"&gt;&lt;/i&gt; Sharing templates and community efforts

&lt;i class="fas fa-question"&gt;&lt;/i&gt; Feature requests

&lt;i class="fas fa-question"&gt;&lt;/i&gt; Bioconductor interests and efforts
]

.youreup[
.center[
You're up !
]
]

- message in the &lt;i class="far fa-comment-alt"&gt;&lt;/i&gt; Text chat
- take the &lt;i class="fas fa-microphone"&gt;&lt;/i&gt; mic
- make a &lt;i class="fas fa-chart-bar"&gt;&lt;/i&gt; poll.
- edit the &lt;i class="fab fa-google-drive"&gt;&lt;/i&gt; [Google Docs](https://docs.google.com/document/d/1ok2H9HVoa8c-m-VcwdQM0ccOuMTmESKCzaaWHb6trwE/edit?usp=sharing)
- join the [#reproducibility](https://community-bioc.slack.com/archives/C0185MUQ8TA) &lt;i class="fab fa-slack"&gt;&lt;/i&gt; Slack channel
- &lt;i class="fab fa-github"&gt;&lt;/i&gt;
Contribute to the [gitbook](https://github.com/kevinrue/ORCA)

---

# References

.small-p[
&lt;a name=bib-mark2020pipelines&gt;&lt;/a&gt;[Marx, V.](#cite-mark2020pipelines)
(2020). "When computational pipelines go 'clank'". In: _Nat Methods_
17.7, pp. 659-662. ISSN: 1548-7105 (Electronic) 1548-7091 (Linking).
DOI:
[10.1038/s41592-020-0886-9](https://doi.org/10.1038%2Fs41592-020-0886-9).
URL:
[https://www.ncbi.nlm.nih.gov/pubmed/32601427](https://www.ncbi.nlm.nih.gov/pubmed/32601427).
]

.thanks[
.center[
Thank you!
]
]

&lt;img src="https://raw.githubusercontent.com/Bioconductor/BiocStickers/f792bce2617124b9e6c4fde55b04909125f8db3f/boards/CAB/CAB_logo_drawing.png" height="400px" style="display: block; margin: auto;" /&gt;
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
